File: mdix-java/src/main/java/com/midmanstudio/dixscript/MdixQuery.java
// MdixQuery.java
package com.midmanstudio.dixscript;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A chainable, LINQ-style query over a set of {@link MdixValue}s. Mirrors
 * Rust's {@code dixscript::Runtime::DixQuery} method for method — build one
 * from {@link Database#query(String)} / {@link Database#queryMany(String)},
 * then chain filter/sort/group operations before a terminal read.
 * <p>
 * Every intermediate method (other than the terminal reads at the bottom)
 * returns a new {@code MdixQuery} rather than mutating in place, same as
 * the Rust original — safe to branch a query into two different chains
 * from a shared prefix.
 * <p>
 * For anything this doesn't cover, {@link #stream()} drops down to a plain
 * {@code Stream<MdixValue>} for standard Java stream operations.
 * <pre>{@code
 * List<String> highPriorityNames = db.query("tasks")
 *     .where_(v -> v.field("priority").asLong() != null && v.field("priority").asLong() == 3)
 *     .select(v -> v.field("name").asString());
 *
 * long activeCount = db.query("events")
 *     .where_(v -> "ACTIVE".equals(v.field("kind").asString()))
 *     .count();
 * }</pre>
 */
public final class MdixQuery {

    private final List<MdixValue> items;

    MdixQuery(List<MdixValue> items) {
        this.items = items;
    }

    // ── filtering / projection ───────────────────────────────────────────────

    /** Keeps only elements matching {@code predicate}. (LINQ {@code Where}) */
    public MdixQuery where_(Predicate<MdixValue> predicate) {
        return new MdixQuery(items.stream().filter(predicate).collect(Collectors.toList()));
    }

    /** Keeps only elements whose {@code field} equals {@code value}. Shorthand for the common equality-filter case. */
    public MdixQuery whereFieldEquals(String field, MdixValue value) {
        return where_(v -> v.field(field).equals(value));
    }

    /** Discards the first {@code n} elements. (LINQ {@code Skip}) */
    public MdixQuery skip(int n) {
        if (n >= items.size()) return new MdixQuery(new ArrayList<>());
        return new MdixQuery(new ArrayList<>(items.subList(Math.max(n, 0), items.size())));
    }

    /** Keeps only the first {@code n} elements. (LINQ {@code Take}) */
    public MdixQuery take(int n) {
        return new MdixQuery(new ArrayList<>(items.subList(0, Math.max(0, Math.min(n, items.size())))));
    }

    /** Removes duplicate elements (by {@link MdixValue#equals}), preserving first-seen order. */
    public MdixQuery distinct() {
        List<MdixValue> out = new ArrayList<>();
        for (MdixValue v : items) {
            if (!out.contains(v)) out.add(v);
        }
        return new MdixQuery(out);
    }

    /** Projects each element to a new form. (LINQ {@code Select}) */
    public <T> List<T> select(Function<MdixValue, T> map) {
        return items.stream().map(map).collect(Collectors.toList());
    }

    /** Projects each element through a named field. Shorthand for {@code select(v -> v.field(name))}. */
    public List<MdixValue> selectField(String name) {
        return select(v -> v.field(name));
    }

    // ── ordering ──────────────────────────────────────────────────────────────

    /** Sorts ascending by a derived key. (LINQ {@code OrderBy}) Stable sort. */
    public <K extends Comparable<K>> MdixQuery orderBy(Function<MdixValue, K> key) {
        List<MdixValue> sorted = new ArrayList<>(items);
        sorted.sort(Comparator.comparing(key));
        return new MdixQuery(sorted);
    }

    /** Sorts descending by a derived key. (LINQ {@code OrderByDescending}) Stable sort. */
    public <K extends Comparable<K>> MdixQuery orderByDescending(Function<MdixValue, K> key) {
        List<MdixValue> sorted = new ArrayList<>(items);
        sorted.sort(Comparator.comparing(key).reversed());
        return new MdixQuery(sorted);
    }

    // ── grouping ──────────────────────────────────────────────────────────────

    /** Groups elements by a derived key, preserving first-seen key order. (LINQ {@code GroupBy}) */
    public <K> Map<K, List<MdixValue>> groupBy(Function<MdixValue, K> key) {
        Map<K, List<MdixValue>> groups = new LinkedHashMap<>();
        for (MdixValue v : items) {
            groups.computeIfAbsent(key.apply(v), k -> new ArrayList<>()).add(v);
        }
        return groups;
    }

    // ── terminal predicates / aggregates ────────────────────────────────────

    public boolean any(Predicate<MdixValue> predicate) { return items.stream().anyMatch(predicate); }

    public boolean all(Predicate<MdixValue> predicate) { return items.stream().allMatch(predicate); }

    public int count() { return items.size(); }

    public boolean isEmpty() { return items.isEmpty(); }

    public MdixValue first() { return items.isEmpty() ? null : items.get(0); }

    public MdixValue firstOr(MdixValue fallback) { return items.isEmpty() ? fallback : items.get(0); }

    public MdixValue last() { return items.isEmpty() ? null : items.get(items.size() - 1); }

    public MdixValue nth(int index) { return (index < 0 || index >= items.size()) ? null : items.get(index); }

    /** Sum of every element's numeric value, widened to {@code long}. Non-numeric elements contribute nothing. */
    public long sumInt() {
        long sum = 0;
        for (MdixValue v : items) {
            Long l = v.asLong();
            if (l != null) sum += l;
        }
        return sum;
    }

    /** Sum of every element's numeric value, widened to {@code double}. Non-numeric elements contribute nothing. */
    public double sumFloat() {
        double sum = 0;
        for (MdixValue v : items) {
            Double d = v.asDouble();
            if (d != null) sum += d;
        }
        return sum;
    }

    /** Average of every numeric element's {@code double} value. {@code null} on an empty query or one with no numeric elements. */
    public Double avgFloat() {
        double sum = 0;
        int n = 0;
        for (MdixValue v : items) {
            Double d = v.asDouble();
            if (d != null) { sum += d; n++; }
        }
        return n == 0 ? null : sum / n;
    }

    public <K extends Comparable<K>> MdixValue minByKey(Function<MdixValue, K> key) {
        return items.stream().min(Comparator.comparing(key)).orElse(null);
    }

    public <K extends Comparable<K>> MdixValue maxByKey(Function<MdixValue, K> key) {
        return items.stream().max(Comparator.comparing(key)).orElse(null);
    }

    /** The current result set as a new, independent {@code List}. */
    public List<MdixValue> toList() { return new ArrayList<>(items); }

    /** Drops down to a plain {@code Stream<MdixValue>} for anything not covered above. */
    public Stream<MdixValue> stream() { return items.stream(); }
}

File: mdix-php/src/Mdix/MdixQuery.php
<?php
declare(strict_types=1);

namespace MidManStudio\Mdix;

/**
 * A chainable, LINQ/Stream-style query over decoded DixScript values.
 *
 * Build one from MdixDatabase::query() / MdixDatabase::queryMany(), then
 * chain filter/sort/group operations before a terminal read. Every
 * intermediate method returns a new MdixQuery rather than mutating in
 * place — safe to branch a query into two different chains from a shared
 * prefix.
 *
 * Deliberately built on plain PHP arrays (from json_decode), not a custom
 * value-wrapper type the way the Java/C++ bindings need one — PHP arrays
 * are already the dynamic, freely-indexable structure those bindings had
 * to build from scratch, so there's nothing to add here beyond the
 * fluent chaining itself.
 *
 *   $highPriority = $db->query('tasks')
 *       ->where(fn($t) => ($t['priority'] ?? 0) === 3)
 *       ->orderByDescending(fn($t) => $t['priority'])
 *       ->select(fn($t) => $t['name']);
 *
 * For anything this doesn't cover, toArray() drops back to a plain PHP
 * array for array_map/array_filter/usort etc.
 */
final class MdixQuery
{
    /** @param array<int,mixed> $items */
    public function __construct(private readonly array $items)
    {
    }

    // ── filtering / projection ───────────────────────────────────────────────

    /** Keeps only elements matching $predicate. @param callable(mixed): bool $predicate */
    public function where(callable $predicate): self
    {
        return new self(\array_values(\array_filter($this->items, $predicate)));
    }

    /** Keeps only elements whose $field equals $value (assumes array-shaped elements). */
    public function whereFieldEquals(string $field, mixed $value): self
    {
        return $this->where(
            fn (mixed $item): bool => \is_array($item) && (($item[$field] ?? null) === $value)
        );
    }

    /** Discards the first $n elements. */
    public function skip(int $n): self
    {
        return new self(\array_slice($this->items, \max($n, 0)));
    }

    /** Keeps only the first $n elements. */
    public function take(int $n): self
    {
        return new self(\array_slice($this->items, 0, \max($n, 0)));
    }

    /** Removes duplicate elements (compared by value, not identity), preserving first-seen order. */
    public function distinct(): self
    {
        $seen = [];
        $out  = [];
        foreach ($this->items as $item) {
            $key = \serialize($item);
            if (!isset($seen[$key])) {
                $seen[$key] = true;
                $out[]      = $item;
            }
        }
        return new self($out);
    }

    /**
     * Projects each element to a new form.
     * @param callable(mixed): mixed $map
     * @return array<int,mixed>
     */
    public function select(callable $map): array
    {
        return \array_map($map, $this->items);
    }

    /**
     * Projects each element through a named field (assumes array-shaped elements).
     * @return array<int,mixed>
     */
    public function selectField(string $field): array
    {
        return $this->select(fn (mixed $item): mixed => \is_array($item) ? ($item[$field] ?? null) : null);
    }

    // ── ordering ──────────────────────────────────────────────────────────────

    /** Sorts ascending by a derived key. Stable (PHP's usort is stable since 8.0). @param callable(mixed): mixed $key */
    public function orderBy(callable $key): self
    {
        $items = $this->items;
        \usort($items, fn (mixed $a, mixed $b): int => $key($a) <=> $key($b));
        return new self($items);
    }

    /** Sorts descending by a derived key. @param callable(mixed): mixed $key */
    public function orderByDescending(callable $key): self
    {
        $items = $this->items;
        \usort($items, fn (mixed $a, mixed $b): int => $key($b) <=> $key($a));
        return new self($items);
    }

    // ── grouping ──────────────────────────────────────────────────────────────

    /**
     * Groups elements by a derived key, preserving first-seen key order.
     * @param callable(mixed): array-key $key
     * @return array<array-key, array<int,mixed>>
     */
    public function groupBy(callable $key): array
    {
        $groups = [];
        foreach ($this->items as $item) {
            $groups[$key($item)][] = $item;
        }
        return $groups;
    }

    // ── terminal predicates / aggregates ────────────────────────────────────

    /** @param callable(mixed): bool $predicate */
    public function any(callable $predicate): bool
    {
        foreach ($this->items as $item) {
            if ($predicate($item)) {
                return true;
            }
        }
        return false;
    }

    /** @param callable(mixed): bool $predicate */
    public function all(callable $predicate): bool
    {
        foreach ($this->items as $item) {
            if (!$predicate($item)) {
                return false;
            }
        }
        return true;
    }

    public function count(): int
    {
        return \count($this->items);
    }

    public function isEmpty(): bool
    {
        return $this->items === [];
    }

    public function first(): mixed
    {
        return $this->items[0] ?? null;
    }

    public function firstOr(mixed $fallback): mixed
    {
        return $this->items === [] ? $fallback : $this->items[0];
    }

    public function last(): mixed
    {
        return $this->items === [] ? null : $this->items[\count($this->items) - 1];
    }

    public function nth(int $index): mixed
    {
        return $this->items[$index] ?? null;
    }

    /** Sum of every numeric element, widened to int. Non-numeric elements contribute nothing. */
    public function sumInt(): int
    {
        return (int) \array_sum(\array_filter($this->items, \is_numeric(...)));
    }

    /** Sum of every numeric element, widened to float. Non-numeric elements contribute nothing. */
    public function sumFloat(): float
    {
        return (float) \array_sum(\array_filter($this->items, \is_numeric(...)));
    }

    /** Average of every numeric element. Null on an empty query or one with no numeric elements. */
    public function avgFloat(): ?float
    {
        $numeric = \array_filter($this->items, \is_numeric(...));
        return $numeric === [] ? null : (float) \array_sum($numeric) / \count($numeric);
    }

    /** @param callable(mixed): mixed $key */
    public function minByKey(callable $key): mixed
    {
        $best = null;
        $bestKey = null;
        $found = false;
        foreach ($this->items as $item) {
            $k = $key($item);
            if (!$found || $k < $bestKey) {
                $best = $item;
                $bestKey = $k;
                $found = true;
            }
        }
        return $best;
    }

    /** @param callable(mixed): mixed $key */
    public function maxByKey(callable $key): mixed
    {
        $best = null;
        $bestKey = null;
        $found = false;
        foreach ($this->items as $item) {
            $k = $key($item);
            if (!$found || $k > $bestKey) {
                $best = $item;
                $bestKey = $k;
                $found = true;
            }
        }
        return $best;
    }

    /** @return array<int,mixed> */
    public function toArray(): array
    {
        return $this->items;
    }
}
